File: packages/server/src/parsers/codex.ts
import { readFileSync } from 'fs';
import { basename } from 'path';
import type {
  CodexEvent,
  CodexSessionMeta,
  CodexEventMsg,
  CodexTokenUsage,
  CodexResponseItem,
  CodexTurnContext,
  SessionSummary,
  SessionDetail,
  Message,
  ToolCall,
  ToolResult,
  SessionStats,
  ToolUsageSummary,
} from '../types/index.js';
import { calculateCost } from '../pricing.js';

export function parseCodexSessionFile(filePath: string): SessionDetail | null {
  try {
    const content = readFileSync(filePath, 'utf-8');
    const lines = content.trim().split('\n').filter(line => line.trim());

    if (lines.length === 0) {
      return null;
    }

    const events: CodexEvent[] = [];
    for (const line of lines) {
      try {
        events.push(JSON.parse(line));
      } catch {
        continue;
      }
    }

    if (events.length === 0) {
      return null;
    }

    // Extract session metadata
    const metaEvent = events.find(e => e.type === 'session_meta');
    const meta = metaEvent?.payload as CodexSessionMeta | undefined;
    const sessionId = meta?.id || basename(filePath, '.jsonl');
    const projectPath = meta?.cwd || '';
    const project = projectPath ? basename(projectPath) : basename(filePath, '.jsonl');

    // Extract model from turn_context
    let model: string | undefined;
    for (const event of events) {
      if (event.type === 'turn_context') {
        const ctx = event.payload as CodexTurnContext;
        if (ctx.model) {
          model = ctx.model;
          break;
        }
      }
    }

    // Build tool call / result maps from response_item events (first pass)
    const toolCallsByCallId = new Map<string, { name: string; args: Record<string, unknown> }>();
    const toolResultsByCallId = new Map<string, ToolResult>();
    const toolUsageMap = new Map<string, { count: number; successes: number }>();

    for (const event of events) {
      if (event.type !== 'response_item') continue;
      const item = event.payload as CodexResponseItem;

      if (item.type === 'function_call' || item.type === 'custom_tool_call') {
        const callId = item.call_id || '';
        const agentName = item.name || 'unknown';
        let args: Record<string, unknown> = {};
        const rawArgs = item.type === 'function_call' ? item.arguments : item.input;
        if (rawArgs) {
          try {
            args = JSON.parse(rawArgs);
          } catch {
            args = { raw: rawArgs };
          }
        }
        // function_call items represent subagent invocations; the display name
        // is 'subagent' and the actual function name is stored as agentName.
        toolCallsByCallId.set(callId, { name: 'subagent', args: { agentName, ...args } });
      }

      if (item.type === 'function_call_output' || item.type === 'custom_tool_call_output') {
        const callId = item.call_id || '';
        const toolName = toolCallsByCallId.get(callId)?.name || 'unknown';
        const rawOutput = item.output;
        const outputStr =
          rawOutput == null ? '' :
          typeof rawOutput === 'string' ? rawOutput :
          JSON.stringify(rawOutput, null, 2);
        toolResultsByCallId.set(callId, {
          toolCallId: callId,
          success: true,
          content: outputStr,
        });
        // function_call tools are tracked as 'subagent' in the usage map
        const usageKey = toolName === 'subagent' ? 'subagent' : toolName;
        const existing = toolUsageMap.get(usageKey) || { count: 0, successes: 0 };
        existing.count++;
        existing.successes++;
        toolUsageMap.set(usageKey, existing);
      }

      // tool_search_call: register as a named tool call
      if (item.type === 'tool_search_call' && item.call_id) {
        let args: Record<string, unknown> = {};
        if (item.arguments) {
          try { args = JSON.parse(item.arguments); } catch { args = { raw: item.arguments }; }
        } else if (item.action) {
          args = item.action as unknown as Record<string, unknown>;
        }
        toolCallsByCallId.set(item.call_id, { name: 'tool_search', args });
      }
      if (item.type === 'tool_search_output' && item.call_id) {
        const rawOutput = item.output;
        const outputStr =
          rawOutput == null ? '' :
          typeof rawOutput === 'string' ? rawOutput :
          JSON.stringify(rawOutput, null, 2);
        toolResultsByCallId.set(item.call_id, {
          toolCallId: item.call_id,
          success: true,
          content: outputStr,
        });
        const existing = toolUsageMap.get('tool_search') || { count: 0, successes: 0 };
        existing.count++;
        existing.successes++;
        toolUsageMap.set('tool_search', existing);
      }

      // web_search_call: track usage (no call_id, use action.query as key)
      if (item.type === 'web_search_call' && item.status === 'completed') {
        const existing = toolUsageMap.get('web_search') || { count: 0, successes: 0 };
        existing.count++;
        existing.successes++;
        toolUsageMap.set('web_search', existing);
      }
    }

    // Track apply_patch usage from patch_apply_end events
    for (const event of events) {
      if (event.type !== 'event_msg') continue;
      const msg = event.payload as CodexEventMsg;
      if (msg.type === 'patch_apply_end' && msg.call_id) {
        const existing = toolUsageMap.get('apply_patch') || { count: 0, successes: 0 };
        existing.count++;
        if (msg.success !== false) existing.successes++;
        toolUsageMap.set('apply_patch', existing);
      }
    }

    // Collect token_count events indexed by position for lookahead pairing
    const tokenCountsByIndex = new Map<number, CodexTokenUsage>();
    let finalTotalUsage: CodexTokenUsage | undefined;

    for (let i = 0; i < events.length; i++) {
      const event = events[i];
      if (event.type !== 'event_msg') continue;
      const msg = event.payload as CodexEventMsg;
      if (msg.type === 'token_count' && msg.info?.last_token_usage) {
        tokenCountsByIndex.set(i, msg.info.last_token_usage);
        if (msg.info.total_token_usage) {
          finalTotalUsage = msg.info.total_token_usage;
        }
      }
    }

    // XML wrappers injected by Codex — not real user/system content, skip entirely.
    const IGNORED_PREFIXES = ['<environment_context'];
    // These injections are not real user input but are shown as system messages.
    const SYSTEM_INJECTION_PREFIXES = [
      '<turn_aborted',
      '<permissions',
      '<agent_response',
      '<user_shell_command',
      '<subagent_notification',
    ];
    function shouldIgnore(text: string): boolean {
      const t = text.trimStart();
      return IGNORED_PREFIXES.some(p => t.startsWith(p));
    }
    function isSystemInjection(text: string): boolean {
      const t = text.trimStart();
      return SYSTEM_INJECTION_PREFIXES.some(p => t.startsWith(p));
    }

    function extractText(content: Array<{ type: string; text?: string }>): string {
      return content.map(c => c.text ?? '').join('').trim();
    }

    // Build messages from response_item events (primary source of truth).
    // event_msg events are only used for duration and session-level token totals.
    //
    // Ordering challenge: commentary assistant messages appear BEFORE the
    // function_call items they narrate.  We therefore buffer each assistant
    // message and flush it (with its following tool calls/results) only when
    // the NEXT assistant message or user message arrives.
    //
    //   commentary msg  ← buffer this
    //   function_call × N          → collect into bufToolCalls
    //   function_call_output × N   → collect into bufToolResults
    //   next msg arrives → flush: emit buffered msg (+toolCalls), then tool results

    const messages: Message[] = [];
    let totalDuration = 0;
    const inputPerMessage: number[] = [];
    const outputPerMessage: number[] = [];
    const cumulativeTokens: number[] = [];
    let cumulativeTotal = 0;

    // Buffered assistant message waiting to be flushed
    let bufMsg: Message | null = null;
    // Tool calls that belong to bufMsg (function_call items following it)
    let bufToolCalls: ToolCall[] = [];
    // Tool result messages that follow the tool calls above
    let bufToolResults: Message[] = [];

    // Track hierarchy: assistant → child of last user, tool → child of its assistant
    let lastUserOrSystemId: string | null = null;
    let lastAssistantId: string | null = null;

    function flushBuffer(): void {
      if (!bufMsg) return;
      if (bufToolCalls.length > 0) bufMsg.toolCalls = bufToolCalls;
      // Set tool results' parentId to their triggering assistant message
      for (const tr of bufToolResults) {
        tr.parentId = bufMsg.id;
      }
      messages.push(bufMsg);
      lastAssistantId = bufMsg.id;
      messages.push(...bufToolResults);
      bufMsg = null;
      bufToolCalls = [];
      bufToolResults = [];
    }

    for (let i = 0; i < events.length; i++) {
      const event = events[i];

      if (event.type === 'response_item') {
        const item = event.payload as CodexResponseItem;

        if (item.type === 'message') {
          const content = item.content || [];
          const text = extractText(content);

          if (item.role === 'developer') {
            if (!shouldIgnore(text)) {
              flushBuffer();
              const id = `system-${event.timestamp}`;
              lastUserOrSystemId = id;
              messages.push({
                id,
                parentId: null,
                role: 'system',
                content: text,
                timestamp: event.timestamp,
              });
            }

          } else if (item.role === 'user') {
            if (shouldIgnore(text)) {
              // drop silently
            } else if (isSystemInjection(text)) {
              flushBuffer();
              const id = `system-${event.timestamp}`;
              lastUserOrSystemId = id;
              messages.push({
                id,
                parentId: null,
                role: 'system',
                content: text,
                timestamp: event.timestamp,
              });
            } else if (text) {
              flushBuffer();
              const id = `user-${event.timestamp}`;
              lastUserOrSystemId = id;
              messages.push({
                id,
                parentId: null,
                role: 'user',
                content: text,
                timestamp: event.timestamp,
              });
            }

          } else if (item.role === 'assistant') {
            // Flush the previous buffered message before starting a new one
            flushBuffer();

            // Look ahead for the nearest token_count (final_answer only)
            let tokens: Message['tokens'] | undefined;
            if (item.phase !== 'commentary') {
              for (let j = i + 1; j < Math.min(events.length, i + 20); j++) {
                const tc = tokenCountsByIndex.get(j);
                if (tc) {
                  const msgCost = calculateCost(
                    { input: tc.input_tokens, output: tc.output_tokens, cacheRead: tc.cached_input_tokens },
                    model
                  );
                  tokens = {
                    input: tc.input_tokens,
                    output: tc.output_tokens,
                    cacheRead: tc.cached_input_tokens,
                    cost: msgCost,
                  };
                  inputPerMessage.push(tc.input_tokens);
                  outputPerMessage.push(tc.output_tokens);
                  cumulativeTotal += tc.input_tokens + tc.output_tokens;
                  cumulativeTokens.push(cumulativeTotal);
                  break;
                }
              }
            }

            bufMsg = {
              id: `assistant-${event.timestamp}`,
              parentId: lastUserOrSystemId,
              role: 'assistant',
              content: text,
              timestamp: event.timestamp,
              model,
              tokens,
            };
          }

        } else if (item.type === 'function_call' || item.type === 'custom_tool_call') {
          const callId = item.call_id || '';
          const toolInfo = toolCallsByCallId.get(callId);
          if (toolInfo) {
            bufToolCalls.push({ id: callId, name: toolInfo.name, arguments: toolInfo.args });
          }

        } else if (item.type === 'tool_search_call' && item.call_id) {
          const toolInfo = toolCallsByCallId.get(item.call_id);
          if (toolInfo) {
            bufToolCalls.push({ id: item.call_id, name: toolInfo.name, arguments: toolInfo.args });
          }

        } else if (item.type === 'web_search_call') {
          const query = item.action?.query || item.action?.queries?.[0] || '';
          bufToolCalls.push({
            id: `ws-${event.timestamp}`,
            name: 'web_search',
            arguments: { query },
          });

        } else if (item.type === 'function_call_output' || item.type === 'custom_tool_call_output') {
          const callId = item.call_id || '';
          const result = toolResultsByCallId.get(callId);
          const toolInfo = toolCallsByCallId.get(callId);
          if (result) {
            bufToolResults.push({
              id: `${callId}-result`,
              parentId: null,
              role: 'tool',
              content: result.content,
              timestamp: event.timestamp,
              toolResult: result,
              toolCalls: toolInfo
                ? [{ id: callId, name: toolInfo.name, arguments: toolInfo.args }]
                : undefined,
            });
          }

        } else if (item.type === 'tool_search_output' && item.call_id) {
          const result = toolResultsByCallId.get(item.call_id);
          const toolInfo = toolCallsByCallId.get(item.call_id);
          if (result) {
            bufToolResults.push({
              id: `${item.call_id}-result`,
              parentId: null,
              role: 'tool',
              content: result.content,
              timestamp: event.timestamp,
              toolResult: result,
              toolCalls: toolInfo
                ? [{ id: item.call_id, name: toolInfo.name, arguments: toolInfo.args }]
                : undefined,
            });
          }
        }
        // reasoning: skip (internal chain-of-thought)
        continue;
      }

      if (event.type !== 'event_msg') continue;
      const msg = event.payload as CodexEventMsg;

      if (msg.type === 'task_complete' && msg.duration_ms) {
        totalDuration += msg.duration_ms;
      }
    }

    // Flush any remaining buffered message at end of stream
    flushBuffer();

    // Build tool usage summary
    const toolUsage: ToolUsageSummary[] = Array.from(toolUsageMap.entries()).map(
      ([name, { count, successes }]) => ({
        name,
        count,
        successRate: count > 0 ? successes / count : 0,
      })
    );

    const userMessages = messages.filter(m => m.role === 'user').length;
    const assistantMessages = messages.filter(m => m.role === 'assistant').length;

    const tokenStats = finalTotalUsage
      ? {
          totalInput: finalTotalUsage.input_tokens,
          totalOutput: finalTotalUsage.output_tokens,
          totalCacheRead: finalTotalUsage.cached_input_tokens,
          totalCacheCreation: 0,
          totalCost: calculateCost(
            { input: finalTotalUsage.input_tokens, output: finalTotalUsage.output_tokens, cacheRead: finalTotalUsage.cached_input_tokens },
            model
          ),
          inputPerMessage,
          outputPerMessage,
          cumulativeTokens,
        }
      : undefined;

    const totalTokens = finalTotalUsage
      ? finalTotalUsage.input_tokens + finalTotalUsage.output_tokens
      : undefined;

    const stats: SessionStats = {
      messageCount: messages.length,
      userMessages,
      assistantMessages,
      tokens: tokenStats,
      tools: toolUsage.map(t => ({ name: t.name, count: t.count, successRate: t.successRate })),
      duration: totalDuration,
    };

    const startTime = messages[0]?.timestamp || meta?.timestamp || new Date().toISOString();
    const lastActivity = messages[messages.length - 1]?.timestamp || startTime;

    return {
      id: sessionId,
      source: 'codex',
      project,
      projectPath,
      startTime,
      lastActivity,
      messageCount: messages.length,
      totalTokens,
      model,
      messages,
      stats,
      toolUsage,
    };
  } catch (error) {
    console.error(`Error parsing Codex session file ${filePath}:`, error);
    return null;
  }
}

export function getCodexSessionSummary(detail: SessionDetail): SessionSummary {
  return {
    id: detail.id,
    source: detail.source,
    project: detail.project,
    projectPath: detail.projectPath,
    startTime: detail.startTime,
    lastActivity: detail.lastActivity,
    messageCount: detail.messageCount,
    totalTokens: detail.totalTokens,
    model: detail.model,
  };
}

File: server-go/parsers/codex.go
package parsers

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"agent-session-viewer/pricing"
	"agent-session-viewer/types"
)

// ──────────────────────────────────────────────────────────
// Raw Codex event types
// ──────────────────────────────────────────────────────────

type codexTokenUsage struct {
	InputTokens           int `json:"input_tokens"`
	CachedInputTokens     int `json:"cached_input_tokens"`
	OutputTokens          int `json:"output_tokens"`
	ReasoningOutputTokens int `json:"reasoning_output_tokens,omitempty"`
	TotalTokens           int `json:"total_tokens"`
}

type codexSessionMeta struct {
	ID            string `json:"id"`
	Timestamp     string `json:"timestamp"`
	CWD           string `json:"cwd"`
	Originator    string `json:"originator,omitempty"`
	CLIVersion    string `json:"cli_version,omitempty"`
	ModelProvider string `json:"model_provider,omitempty"`
}

type codexEventMsg struct {
	Type string `json:"type"`
	// user_message / agent_message
	Message string `json:"message,omitempty"`
	Phase   string `json:"phase,omitempty"`
	// task_complete
	DurationMs *int64 `json:"duration_ms,omitempty"`
	// token_count
	Info *struct {
		TotalTokenUsage *codexTokenUsage `json:"total_token_usage,omitempty"`
		LastTokenUsage  *codexTokenUsage `json:"last_token_usage,omitempty"`
	} `json:"info,omitempty"`
	// patch_apply_end
	CallID  string `json:"call_id,omitempty"`
	Success *bool  `json:"success,omitempty"`
}

type codexResponseItem struct {
	Type      string      `json:"type"`
	Name      string      `json:"name,omitempty"`
	Arguments string      `json:"arguments,omitempty"`
	Input     string      `json:"input,omitempty"`
	CallID    string      `json:"call_id,omitempty"`
	Output    interface{} `json:"output,omitempty"`
	Role      string      `json:"role,omitempty"`
	Content   []struct {
		Type string `json:"type"`
		Text string `json:"text,omitempty"`
	} `json:"content,omitempty"`
}

type codexTurnContext struct {
	TurnID string `json:"turn_id"`
	CWD    string `json:"cwd"`
	Model  string `json:"model,omitempty"`
	Effort string `json:"effort,omitempty"`
}

type codexEvent struct {
	Timestamp string          `json:"timestamp"`
	Type      string          `json:"type"`
	Payload   json.RawMessage `json:"payload"`
}

// ──────────────────────────────────────────────────────────
// Parser
// ──────────────────────────────────────────────────────────

func ParseCodexSessionFile(filePath string) *types.SessionDetail {
	events := readCodexEvents(filePath)
	if len(events) == 0 {
		return nil
	}

	// Extract metadata
	var meta *codexSessionMeta
	for _, ev := range events {
		if ev.Type == "session_meta" {
			var m codexSessionMeta
			if json.Unmarshal(ev.Payload, &m) == nil {
				meta = &m
				break
			}
		}
	}

	sessionID := filepath.Base(filePath)
	if meta != nil && meta.ID != "" {
		sessionID = meta.ID
	}
	sessionID = strings.TrimSuffix(sessionID, ".jsonl")

	projectPath := ""
	if meta != nil {
		projectPath = meta.CWD
	}
	project := filepath.Base(projectPath)
	if project == "" || project == "." {
		project = strings.TrimSuffix(filepath.Base(filePath), ".jsonl")
	}

	// Extract model from first turn_context
	model := ""
	for _, ev := range events {
		if ev.Type == "turn_context" {
			var tc codexTurnContext
			if json.Unmarshal(ev.Payload, &tc) == nil && tc.Model != "" {
				model = tc.Model
				break
			}
		}
	}

	// Build tool call/result maps
	toolCallsByCallID := map[string]struct {
		name string
		args map[string]interface{}
	}{}
	toolResultsByCallID := map[string]types.ToolResult{}
	toolUsageMap := map[string]struct{ count, successes int }{}

	for _, ev := range events {
		if ev.Type != "response_item" {
			continue
		}
		var item codexResponseItem
		if json.Unmarshal(ev.Payload, &item) != nil {
			continue
		}
		if item.Type == "function_call" || item.Type == "custom_tool_call" {
			callID := item.CallID
			name := item.Name
			if name == "" {
				name = "unknown"
			}
			rawArgs := item.Arguments
			if rawArgs == "" {
				rawArgs = item.Input
			}
			args := map[string]interface{}{}
			if rawArgs != "" {
				if json.Unmarshal([]byte(rawArgs), &args) != nil {
					args = map[string]interface{}{"raw": rawArgs}
				}
			}
			toolCallsByCallID[callID] = struct {
				name string
				args map[string]interface{}
			}{name, args}
		}
		if item.Type == "function_call_output" || item.Type == "custom_tool_call_output" {
			callID := item.CallID
			toolName := "unknown"
			if tc, ok := toolCallsByCallID[callID]; ok {
				toolName = tc.name
			}
			outputStr := ""
			if item.Output != nil {
				switch v := item.Output.(type) {
				case string:
					outputStr = v
				default:
					b, _ := json.MarshalIndent(v, "", "  ")
					outputStr = string(b)
				}
			}
			toolResultsByCallID[callID] = types.ToolResult{
				ToolCallID: callID,
				Success:    true,
				Content:    outputStr,
			}
			u := toolUsageMap[toolName]
			u.count++
			u.successes++
			toolUsageMap[toolName] = u
		}
	}

	// Track patch_apply_end as tool usage
	for _, ev := range events {
		if ev.Type != "event_msg" {
			continue
		}
		var msg codexEventMsg
		if json.Unmarshal(ev.Payload, &msg) != nil {
			continue
		}
		if msg.Type == "patch_apply_end" && msg.CallID != "" {
			u := toolUsageMap["apply_patch"]
			u.count++
			if msg.Success == nil || *msg.Success {
				u.successes++
			}
			toolUsageMap["apply_patch"] = u
		}
	}

	// Collect token_count events indexed by event position
	tokenCountsByIdx := map[int]*codexTokenUsage{}
	var finalTotalUsage *codexTokenUsage
	for i, ev := range events {
		if ev.Type != "event_msg" {
			continue
		}
		var msg codexEventMsg
		if json.Unmarshal(ev.Payload, &msg) != nil {
			continue
		}
		if msg.Type == "token_count" && msg.Info != nil {
			if msg.Info.LastTokenUsage != nil {
				tokenCountsByIdx[i] = msg.Info.LastTokenUsage
			}
			if msg.Info.TotalTokenUsage != nil {
				finalTotalUsage = msg.Info.TotalTokenUsage
			}
		}
	}

	// Build messages
	messages := []types.Message{}
	var totalDuration int64
	inputPerMsg := []int{}
	outputPerMsg := []int{}
	cumulativeTokens := []int{}
	cumTotal := 0
	var pendingToolCalls []types.ToolCall

	for i, ev := range events {
		if ev.Type == "response_item" {
			var item codexResponseItem
			if json.Unmarshal(ev.Payload, &item) != nil {
				continue
			}
			if item.Type == "function_call" || item.Type == "custom_tool_call" {
				if tc, ok := toolCallsByCallID[item.CallID]; ok {
					pendingToolCalls = append(pendingToolCalls, types.ToolCall{
						ID:        item.CallID,
						Name:      tc.name,
						Arguments: tc.args,
					})
				}
			} else if item.Type == "function_call_output" || item.Type == "custom_tool_call_output" {
				if r, ok := toolResultsByCallID[item.CallID]; ok {
					messages = append(messages, types.Message{
						ID:         fmt.Sprintf("%s-result", item.CallID),
						ParentID:   nil,
						Role:       "tool",
						Content:    r.Content,
						Timestamp:  ev.Timestamp,
						ToolResult: &r,
					})
				}
			}
			continue
		}

		if ev.Type != "event_msg" {
			continue
		}
		var msg codexEventMsg
		if json.Unmarshal(ev.Payload, &msg) != nil {
			continue
		}

		switch msg.Type {
		case "user_message":
			messages = append(messages, types.Message{
				ID:        fmt.Sprintf("user-%s", ev.Timestamp),
				ParentID:  nil,
				Role:      "user",
				Content:   msg.Message,
				Timestamp: ev.Timestamp,
			})
			pendingToolCalls = nil

		case "agent_message":
			toolCalls := pendingToolCalls
			pendingToolCalls = nil

			var tokens *types.Tokens
			// Look ahead up to 10 events for a token_count
			for j := i + 1; j < len(events) && j < i+10; j++ {
				if tc, ok := tokenCountsByIdx[j]; ok {
					c := pricing.CalculateCost(pricing.TokenCounts{
						Input:     tc.InputTokens,
						Output:    tc.OutputTokens,
						CacheRead: tc.CachedInputTokens,
					}, model)
					tokens = &types.Tokens{
						Input:     tc.InputTokens,
						Output:    tc.OutputTokens,
						CacheRead: intPtr(tc.CachedInputTokens),
						Cost:      &c,
					}
					inputPerMsg = append(inputPerMsg, tc.InputTokens)
					outputPerMsg = append(outputPerMsg, tc.OutputTokens)
					cumTotal += tc.InputTokens + tc.OutputTokens
					cumulativeTokens = append(cumulativeTokens, cumTotal)
					break
				}
			}

			m := types.Message{
				ID:        fmt.Sprintf("agent-%s", ev.Timestamp),
				ParentID:  nil,
				Role:      "assistant",
				Content:   msg.Message,
				Timestamp: ev.Timestamp,
				Tokens:    tokens,
			}
			if model != "" {
				m.Model = &model
			}
			if len(toolCalls) > 0 {
				m.ToolCalls = toolCalls
			}
			messages = append(messages, m)

		case "task_complete":
			if msg.DurationMs != nil {
				totalDuration += *msg.DurationMs
			}
		}
	}

	// Build tool summaries
	var toolUsage []types.ToolUsageSummary
	var toolStats []types.ToolStat
	for name, u := range toolUsageMap {
		sr := 0.0
		if u.count > 0 {
			sr = float64(u.successes) / float64(u.count)
		}
		toolUsage = append(toolUsage, types.ToolUsageSummary{Name: name, Count: u.count, SuccessRate: sr})
		toolStats = append(toolStats, types.ToolStat{Name: name, Count: u.count, SuccessRate: sr})
	}
	sort.Slice(toolUsage, func(i, j int) bool { return toolUsage[i].Count > toolUsage[j].Count })
	sort.Slice(toolStats, func(i, j int) bool { return toolStats[i].Count > toolStats[j].Count })

	userMsgs, assistantMsgs := 0, 0
	for _, m := range messages {
		if m.Role == "user" {
			userMsgs++
		} else if m.Role == "assistant" {
			assistantMsgs++
		}
	}

	var tokenStats *types.TokenStats
	if finalTotalUsage != nil {
		c := pricing.CalculateCost(pricing.TokenCounts{
			Input:     finalTotalUsage.InputTokens,
			Output:    finalTotalUsage.OutputTokens,
			CacheRead: finalTotalUsage.CachedInputTokens,
		}, model)
		tokenStats = &types.TokenStats{
			TotalInput:       finalTotalUsage.InputTokens,
			TotalOutput:      finalTotalUsage.OutputTokens,
			TotalCacheRead:   finalTotalUsage.CachedInputTokens,
			TotalCost:        &c,
			InputPerMessage:  inputPerMsg,
			OutputPerMessage: outputPerMsg,
			CumulativeTokens: cumulativeTokens,
		}
	}

	stats := types.SessionStats{
		MessageCount:      len(messages),
		UserMessages:      userMsgs,
		AssistantMessages: assistantMsgs,
		Tokens:            tokenStats,
		Tools:             toolStats,
		Duration:          totalDuration,
	}

	startTime := time.Now().UTC().Format(time.RFC3339)
	if meta != nil && meta.Timestamp != "" {
		startTime = meta.Timestamp
	}
	lastActivity := startTime
	if len(messages) > 0 {
		startTime = messages[0].Timestamp
		lastActivity = messages[len(messages)-1].Timestamp
	}

	var totalTok *int
	if finalTotalUsage != nil {
		t := finalTotalUsage.InputTokens + finalTotalUsage.OutputTokens
		totalTok = &t
	}
	var modelPtr *string
	if model != "" {
		modelPtr = &model
	}

	return &types.SessionDetail{
		SessionSummary: types.SessionSummary{
			ID:           sessionID,
			Source:       types.SourceCodex,
			Project:      project,
			ProjectPath:  projectPath,
			StartTime:    startTime,
			LastActivity: lastActivity,
			MessageCount: len(messages),
			TotalTokens:  totalTok,
			Model:        modelPtr,
		},
		Messages:  messages,
		Stats:     stats,
		ToolUsage: toolUsage,
	}
}

func readCodexEvents(filePath string) []codexEvent {
	f, err := os.Open(filePath)
	if err != nil {
		return nil
	}
	defer f.Close()

	var events []codexEvent
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 4*1024*1024), 4*1024*1024)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		var ev codexEvent
		if json.Unmarshal([]byte(line), &ev) == nil {
			events = append(events, ev)
		}
	}
	return events
}

func intPtr(n int) *int { return &n }
